pydantic_graph/pydantic_graph/graph.py

```diff
@@ -510,3 +505,109 @@ def _infer_name(self, function_frame: types.FrameType | None) -> None:
+
+        self._next_node = first_node
+        self._started: bool = False
+        self._result: End[RunEndT] | None = None
+
+    @property
+    def is_ended(self):
+        return self._result is not None
+
+    @property
+    def result(self) -> RunEndT:
+        if self._result is None:
+            if self._started:
+                raise exceptions.GraphRuntimeError(
+                    'This GraphRun has not yet ended. Continue iterating with `async for` or `GraphRun.next`'
+                    ' to complete the run before accessing the result.'
+                )
+            else:
+                raise exceptions.GraphRuntimeError(
+                    'This GraphRun has not been started. Did you forget to `await` the run?'
+                )
+        return self._result.data
+
+    async def next(
+        self: GraphRun[StateT, DepsT, T], node: BaseNode[StateT, DepsT, T]
+    ) -> BaseNode[StateT, DepsT, T] | End[T]:
+        """Note: this method behaves very similarly to an async generator's `asend` method."""
+        history = self.history
+        state = self.state
+        deps = self.deps
+
+        next_node = await self.graph.next(node, history, state=state, deps=deps, infer_name=False)
+
+        if isinstance(next_node, End):
+            self._result = next_node
+        else:
+            self._next_node = next_node
+        return next_node
+
+    def __await__(self) -> Generator[Any, Any, typing_extensions.Self]:
+        """Run the graph until it ends, and return the final result."""
+
+        async def _run() -> typing_extensions.Self:
+            with self:
+                async for _next_node in self:
+                    pass
+
+                return self
+
+        return _run().__await__()
+
+    def __enter__(self) -> typing_extensions.Self:
+        if self._started:
+            raise exceptions.GraphRuntimeError('A GraphRun can only be started once.')
+
+        if self._auto_instrument and self._span is None:
+            self._span = logfire_api.span('run graph {graph.name}', graph=self.graph)
+
+        if self._span is not None:
+            self._span.__enter__()
+
+        self._started = True
+        return self
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        if self._span is not None:
+            self._span.__exit__(exc_type, exc_val, exc_tb)
+        self._span = None  # make it more obvious if you try to use it after exiting
+
+    def __aiter__(self) -> AsyncIterator[BaseNode[StateT, DepsT, RunEndT] | End[RunEndT]]:
+        return self
+
+    async def __anext__(self) -> BaseNode[StateT, DepsT, RunEndT] | End[RunEndT]:
+        """Use the last returned node as the input to `Graph.next`."""
+        if self._result:
+            raise StopAsyncIteration
+        if not self._started:
+            raise exceptions.GraphRuntimeError(
+                'You must enter the GraphRun as a contextmanager before you can iterate over it.'
+            )
```

Comment:
should we do this check without `next`?

Reply:
I'm not sure I follow this comment, but if you are saying that we should do the same check in `next`, I agree, and I will make that change. If you are saying something else, I don't understand